src/index.js

```diff
@@ -51,9 +56,34 @@ const Reliable = or(
   UTP
 )
 
-const IPFS = or(
+let _IPFS = or(
   and(Reliable, base('ipfs')),
-  WebRTCStar
+  WebRTCStar,
+  base('ipfs')
+)
+
+const _Circuit = or(
+  and(_IPFS, base('p2p-circuit'), _IPFS),
+  and(_IPFS, base('p2p-circuit')),
+  and(base('p2p-circuit'), _IPFS),
+  and(Reliable, base('p2p-circuit')),
+  and(base('p2p-circuit'), Reliable),
+  base('p2p-circuit')
+)
+
+const CircuitRecursive = () => or(
```

Comment:
I think I know what your talking about, we can just do `() => Circuit` when we need recursivity there.